lxd/instance/drivers/driver_qemu.go

```diff
@@ -1600,6 +1600,105 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 	return nil
 }
 
+func (d *qemu) setupSEV(sevOpts *qemuSevOpts, fdFiles *[]*os.File) error {
+	if d.architecture != osarch.ARCH_64BIT_INTEL_X86 {
+		return errors.New("AMD SEV support is only available on x86_64 systems")
+	}
+
+	qemuPath, _, err := d.qemuArchConfig(d.architecture)
+	if err != nil {
+		return err
+	}
+
+	// Get the QEMU features to check if AMD SEV is supported.
+	features, err := d.checkFeatures(d.architecture, qemuPath)
+	if err != nil {
+		return err
+	}
+
+	_, smeFound := features["sme"]
+	sev, sevFound := features["sev"]
+	if !smeFound || !sevFound {
+		return errors.New("AMD SEV is not supported by the host")
+	}
+
+	// Get the SEV guest `cbitpos` and `reducedPhysBits`.
+	sevCapabilities, ok := sev.(qmp.AMDSEVCapabilities)
+	if !ok {
+		return errors.New("Failed to get the guest `sev` capabilities")
+	}
+
+	cbitpos := sevCapabilities.CBitPos
+	reducedPhysBits := sevCapabilities.ReducedPhysBits
+
+	// Write user's dh-cert and session-data to file descriptors.
+	var dhCertFD, sessionDataFD int
+	var dhCertFDOk, sessionDataFDOk bool
+	if d.localConfig["security.sev.session.dh"] != "" {
+		dhCert, err := os.CreateTemp("", "lxd_sev_dh_cert_")
+		if err != nil {
+			return err
+		}
+
+		err = os.Remove(dhCert.Name())
+		if err != nil {
+			return err
+		}
+
+		_, err = dhCert.WriteString(d.localConfig["security.sev.session.dh"])
+		if err != nil {
+			return err
+		}
+
+		dhCertFD = d.addFileDescriptor(fdFiles, dhCert)
+		defer dhCert.Close()
```

Comment:
I'm deleting the `defer dhCert.Close()` and defer `sessionData.Close()` lines but as you can see line 1411, the files are closed with a deferred call. Plus, adding the deferred calls line 1223 would not contain our files that are being added after so I think this should be good.

Reply:
@gabrielmougard mmm, ok, but we still need to have the cleanup code I posted above (in the place I suggested) and then remove the existing defer cleanup code on line 1411. As otherwise if the VM start functions fails to get to line 1411 then the file handles will be left open and leak.

> Plus, adding the deferred calls line 1223 would not contain our files that are being added after so I think this should be good.

Thats not quite right. Whilst true that the file handles are not in fdFiles at the time of the proposed defer, notice that I am not suggesting we use the existing invocation of:

```go
// Ensure passed files are closed after qemu has started.
for _, file := range fdFiles {
    defer func(file *os.File) { _ = file.Close() }(file)
}
```

Which, as you say, would not actually do anything because fdFiles is empty.

I am proposing using:

```go
// Ensure passed files are closed after start has returned (either because qemu has started or on error).
defer func() {
	for _, file := range fdFiles {
		_ = file.Close()
	}
}()
```

Which says instead: "at the end of the function, loop through fdFiles and close them".

Reply:
oh right ! Very subtle approach indeed ! I'm adding this bit then ;)